flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -0,0 +1,317 @@
+
+        if (successfulUpdate) {
+            if (taskExecutionStateTransition.getExecutionState() == ExecutionState.FAILED) {
+                Throwable cause = taskExecutionStateTransition.getError(userCodeClassLoader);
+                handleAnyFailure(cause);
+            }
+        }
+
+        return successfulUpdate;
+    }
+
+    @Override
+    void onGloballyTerminalState(JobStatus globallyTerminalState) {
+        context.goToFinished(ArchivedExecutionGraph.createFrom(getExecutionGraph()));
+    }
+
+    private void handleAnyFailure(Throwable cause) {
+        final Executing.FailureResult failureResult = context.howToHandleFailure(cause);
+
+        if (failureResult.canRestart()) {
+            context.goToRestarting(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getBackoffTime());
+        } else {
+            context.goToFailing(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getFailureCause());
+        }
+    }
+
+    CompletableFuture<String> getOperationCompletionFuture() {
+        return operationCompletionFuture;
+    }
+
+    @Override
+    public void startCheckpointScheduler() {
+        final CheckpointCoordinator coordinator = getExecutionGraph().getCheckpointCoordinator();
+        if (coordinator == null) {
+            if (!getExecutionGraph().getState().isTerminalState()) {
+                // for a streaming job, the checkpoint coordinator is always set (even if periodic
+                // checkpoints are disabled). The only situation where it can be null is when the
+                // job reached a terminal state.
+                throw new IllegalStateException(
+                        "Coordinator is only allowed to be null if we are in a terminal state.");
+            }
+            return;
+        }
+        if (coordinator.isPeriodicCheckpointingConfigured()) {
+            coordinator.startCheckpointScheduler();
+        }
+    }
+
+    @Override
+    public void stopCheckpointScheduler() {
+        final CheckpointCoordinator coordinator = getExecutionGraph().getCheckpointCoordinator();
+        if (coordinator != null) {
+            coordinator.stopCheckpointScheduler();
+        }
+    }
+
+    @Override
+    public CompletableFuture<CompletedCheckpoint> triggerSynchronousSavepoint(
+            boolean terminate, @Nullable String targetLocation) {
+        Preconditions.checkNotNull(
+                getExecutionGraph().getCheckpointCoordinator(),
+                "Checkpoint coordinator must be set for this operation. Is this a streaming job?");
+        return getExecutionGraph()
+                .getCheckpointCoordinator()
+                .triggerSynchronousSavepoint(terminate, targetLocation);
```

Comment:
I don't understand what you are hinting at here? (note that my updated version of this PR won't expose this method as a public method of StopWithSavepoint anymore. Otherwise, the respective future completions are monitored and treated accordingly.)

Reply:
I can't remember either.